### PR TITLE
Refactored to compile with openssl 1.1

### PR DESCRIPTION
**Pull request by urcheon**
Tested all custom chil/hwchrk functions using a nFast PCI nC1002P/nC3022P ver 2.18.13 card.
Both embed and hwcrhk "application" types work.

### CODE REVIEW

File: e_chil.c [hunk 1]
```diff
@@ -224,8 +166,10 @@ static RAND_METHOD hwcrhk_rand = {
 /* Constants used when creating the ENGINE */
 static const char *engine_hwcrhk_id = "chil";
 static const char *engine_hwcrhk_name = "CHIL hardware engine support";
+#  ifndef OPENSSL_NO_DYNAMIC_ENGINE
 /* Compatibility hack, the dynamic library uses this form in the path */
 static const char *engine_hwcrhk_id_alt = "ncipher";
+#  endif
```
Comment (levitte) [inline e_chil.c]: There's no reason to add this guard.  This gets built as a module, no exception.

File: e_chil.c [hunk 2]
```diff
@@ -340,57 +275,132 @@ static HWCryptoHook_InitInfo hwcrhk_globals = {
...
+{
+    /* Copied from eng_[openssl|dyn].c */
+    ENGINE *toadd = engine_chil();
+    if (!toadd)
+        return;
+    ENGINE_add(toadd);
+    ENGINE_free(toadd);
+    ERR_clear_error();
 }
+#  endif
```
Comment (levitte) [inline e_chil.c]: You can remove the whole section of code from line 380.

File: e_chil_err.c [hunk 3]
```diff
@@ -1,55 +1,10 @@
...
- * This product includes cryptographic software written by Eric Young
- * (eay@cryptsoft.com).  This product includes software written by Tim
- * Hudson (tjh@cryptsoft.com).
+/*
+ * Copyright 1995-2016 The OpenSSL Project Authors. All Rights Reserved.
  *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
```
Comment (levitte) [inline e_chil_err.c]: Same comment on the license...

File: e_chil_err.h [hunk 4]
```diff
@@ -1,55 +1,16 @@
...
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/*
+ * NOTE: this file was auto generated by the mkerr.pl script: any changes
+ * made to it will be overwritten when the script next updates this file,
+ * only reason strings will be preserved.
```
Comment (levitte) [inline e_chil_err.h]: That's a point, though...  is it still true or did you hand edit it?

File: e_chil.c [hunk 5]
```diff
@@ -1,67 +1,17 @@
...
- *
- * This product includes cryptographic software written by Eric Young
- * (eay@cryptsoft.com).  This product includes software written by Tim
- * Hudson (tjh@cryptsoft.com).
+ * Copyright 2000-2016 The OpenSSL Project Authors. All Rights Reserved.
  *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
```
[the remark this answers is not in the document]
Comment (urcheon) [reply e_chil.c]: Sorry, reverted

File: e_chil.c [hunk 6]
```diff
@@ -73,6 +23,9 @@
 #endif
 #include <openssl/bn.h>
 
+#ifndef OPENSSL_NO_HW
+# ifndef OPENSSL_NO_HW_CHIL
+
```
[the remark this answers is not in the document]
Comment (urcheon) [reply e_chil.c]: Removed

File: e_chil.c [hunk 7]
```diff
@@ -82,10 +35,16 @@
...
  */
-#include "vendor_defns/hwcryptohook.h"
+#  ifdef FLAT_INC
+#   include "hwcryptohook.h"
+#  else
+#   include "vendor_defns/hwcryptohook.h"
+#  endif
+
+#  define HWCRHK_LIB_NAME "CHIL engine"
+#  include "e_chil_err.c"
```
[the remark this answers is not in the document]
Comment (urcheon) [reply e_chil.c]: Removed

File: e_chil.c (hunk 1 above)
Comment (urcheon) [reply e_chil.c]: Removed

File: e_chil.c (hunk 2 above)
Comment (urcheon) [reply e_chil.c]: Removed

File: e_chil_err.c (hunk 3 above)
Comment (urcheon) [reply e_chil_err.c]: Sorry, reverted

File: e_chil_err.h (hunk 4 above)
Comment (urcheon) [reply e_chil_err.h]: Yes, good point. Spent 5 minutes trying to run mkerr.pl, and gave up.

File: e_chil.c [hunk 8]
```diff
@@ -1,67 +1,17 @@
...
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
  */
 
 #include <stdio.h>
 #include <string.h>
 #include <openssl/crypto.h>
 #include <openssl/pem.h>
-#include <openssl/dso.h>
+#include "internal/dso.h"
```
[the remark this answers is not in the document]
Comment (urcheon) [reply e_chil.c]: I agree, It is a little ugly using OpenSSL internals.

**Comment by levitte** [comment]
Might I suggest a rebase on top of a fresher master?  I just updated the module to use libltdl (because it's convenient) instead of OpenSSL's DSO functions.

**Comment by urcheon** [comment]
Awesome. Rebased.

File: e_chil.c [hunk 9]
```diff
@@ -1287,17 +1554,16 @@ static void hwcrhk_log_message(void *logstr, const char *message)
...
         lstream = *(BIO **)logstr;
     if (lstream) {
         BIO_printf(lstream, "%s\n", message);
     }
-    CRYPTO_w_unlock(CRYPTO_LOCK_BIO);
 }
 
 /*
- * This stuff is needed to compile this ENGINE into a DSO.
+ * This stuff is needed if this ENGINE is being compiled into a
```
Comment (levitte) [inline e_chil.c]: If I may be so picky, not "if".  We *are* building a DSO ("dlopenable" in GNU terms), there's no alternative here.

Comment (urcheon) [reply e_chil.c]: Done

**Pull request merged**